### PR TITLE
adds new authentication flow to RHSSO to delay user login

### PR DESCRIPTION
**Pull request by jjaferson**
# Description

Add a new Authentication Flow that delays users login until they get created on 3scale by the RHMI operator and flagged as created in KeycloakUser CR. This new authentication flow will replace the First Broker login authentication flow in openshift realm of RHSSO. 

https://issues.redhat.com/browse/INTLY-6491

This PR requires other PRs to be merged and e2e test to be implemented
- keycloak client api https://github.com/integr8ly/keycloak-client/pull/7
- keycloak operator downstream https://github.com/integr8ly/keycloak-operator-rhmi/pull/7

Link to the Keycloak authentication delay plugin: https://github.com/integr8ly/authentication-delay-plugin

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Verification steps

1) Install the operator locally against a OSD cluster
2) Add testing IDP  `NUM_ADMIN=5 NUM_REGULAR_USER=5 ./scripts/setup-sso-idp.sh`
3) Login to 3scale with one of the Openshift users via RHSSO
4) Ensure you a redirected to a page that shows the message `Your account is being provisioned` 
5) Verify user is created in RHSSO and gets the attribute `3scale_user_created` set to `true`
6) Verify if user is redirected and authenticated in 3scale.

### CODE REVIEW

File: pkg/products/rhsso/reconciler_test.go [hunk 1]
```diff
@@ -295,55 +303,37 @@ func TestReconciler_reconcileComponents(t *testing.T) {
...
 					Kind:       integreatlyv1alpha1.SchemaGroupVersionKind.Kind,
 					APIVersion: integreatlyv1alpha1.SchemeGroupVersion.String(),
 				},
-			},
-			ExpectedStatus: integreatlyv1alpha1.PhaseCompleted,
-			Recorder:       setupRecorder(),
-			ApiUrl:         "https://serverurl",
-		},
-		{
-			Name: "Test reconcile custom resource returns failed on unsuccessful create",
```
[the remark this answers is not in the document]
Comment (jjaferson) [reply pkg/products/rhsso/reconciler_test.go:326]: Added back

File: pkg/products/rhsso/reconciler_test.go [hunk 2]
```diff
@@ -875,3 +878,203 @@ func getKcr(status keycloak.KeycloakRealmStatus) *keycloak.KeycloakRealm {
...
+				// dummy ones
+				&keycloak.AuthenticationExecutionInfo{
+					Requirement: "REQUIRED",
+					Alias:       "dummy execution",
+				},
+			},
+		},
+	}
+
+	createAuthenticationFlowFunc := func(authFlow keycloakCommon.AuthenticationFlow, realmName string) (string, error) {
```
[the remark this answers is not in the document]
Comment (jjaferson) [reply pkg/products/rhsso/reconciler_test.go:984]: I believe `createAuthenticationFlow` already fails if you attempt to create a flow that exists already, I think it's a good idea to mock the same behavior.

File: pkg/products/rhsso/reconciler.go [hunk 3]
```diff
@@ -433,6 +437,7 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, installation *inte
 	or, err := controllerutil.CreateOrUpdate(ctx, serverClient, kc, func() error {
 		kc.Spec.Extensions = []string{
 			"https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar",
+			"http://github.com/integr8ly/authentication-delay-plugin/releases/download/1.0.1/authdelay.jar",
```
Comment (matskiv) [inline pkg/products/rhsso/reconciler.go:440]:
```suggestion
			"https://github.com/integr8ly/authentication-delay-plugin/releases/download/1.0.1/authdelay.jar",
```

Comment (jjaferson) [reply pkg/products/rhsso/reconciler.go:440]: oops

File: Makefile [hunk 4]
```diff
@@ -140,7 +140,7 @@ test/e2e/prow: export INTEGREATLY_OPERATOR_IMAGE := "${IMAGE_FORMAT}"
 test/e2e/prow: test/e2e
 
 .PHONY: test/e2e
-test/e2e:  export SURF_DEBUG_HEADERS=1
+# test/e2e:  export SURF_DEBUG_HEADERS=1
```
Comment (matskiv) [inline Makefile:143]: This needs to be un-commented.

File: Makefile [hunk 5]
```diff
@@ -151,6 +151,7 @@ test/e2e/local: cluster/cleanup cluster/cleanup/crds cluster/prepare cluster/pre
 
 
 .PHONY: test/functional
+# test/functional:  export SURF_DEBUG_HEADERS=1
```
Comment (matskiv) [inline Makefile:154]: This needs to be un-commented.

**Comment by matskiv** [comment]
I will trigger another e2e test in PROW just to measure stability of the new test.

/test e2e

File: test/resources/openshift_auth.go [hunk 6]
```diff
@@ -137,7 +137,7 @@ func openshiftClientSetup(url, username, password string, client *http.Client, i
 	//sometimes we'll reach an accept permissions page for the user if they haven't accepted these scope requests before.
 	//refactored, this approach assumes that if the redirected page is not the console then it looks for an approve action, previous approach would cause e2e test flakes
 	if strings.Contains(browser.Url().Host, openshiftConsoleSubdomain) {
-		l.Log("Looking for permissions form on page", browser.Url().Host, browser.Body())
+		// l.Log("Looking for permissions form on page", browser.Url().Host, browser.Body())
```
Comment (matskiv) [inline test/resources/openshift_auth.go:140]: Should the changes in this file be reverted before merging?

Comment (jjaferson) [reply test/resources/openshift_auth.go:140]: I'll revert it back, I just commented it out to make it easier to investigate the reason e2e test was not passing in PROW

File: test/common/tests.go [hunk 7]
```diff
@@ -13,6 +13,7 @@ var (
 	HAPPY_PATH_TESTS = []TestCase{
 		// Add all happy path tests to be executed after RHMI installation is completed here
 		{"Verify RHMI Config CRs Successful", TestRHMIConfigCRs},
+		{"Verify auth flow first broker login", TestAuthDelayFirstBrokerLogin},
```
Comment (matskiv) [inline test/common/tests.go:16]: btw: this test has been added to integreatly-test-cases repo as "A16 - Custom first broker login flow" (https://gitlab.cee.redhat.com/integreatly-qe/integreatly-test-cases/blob/master/tests/installation/a16-custom-first-broker-login-flow.md)
We probably need to rename it to match, or at least prepend the name with "A16" code.

**Comment by jjaferson** [comment]
keycloak operator upgrade from 9.0.2 to 9.0.3 (version the includes user attributes to keycloakuser CR) is tested and working as expected.

https://github.com/integr8ly/keycloak-operator-rhmi/commit/3dc90c783b50ee287709d69c44fb8f0662f865ef

**Pull request merged**